include/aws/common/log.h

```diff
@@ -0,0 +1,124 @@
+* on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+* express or implied. See the License for the specific language governing
+* permissions and limitations under the License.
+*/
+
+#include <stdarg.h>
+
+#include <aws/common/memory_pool.h>
+#include <aws/common/linked_list.h>
+
+enum aws_log_level
+{
+    AWS_LOG_LEVEL_OFF = 0, /* No logs will be processed at all. */
+    AWS_LOG_LEVEL_FATAL,
+    AWS_LOG_LEVEL_ERROR,
+    AWS_LOG_LEVEL_WARN,
+    AWS_LOG_LEVEL_INFO,
+    AWS_LOG_LEVEL_DEBUG,
+    AWS_LOG_LEVEL_TRACE
+};
+
+#define AWS_DEFAULT_LOG_LEVEL AWS_LOG_LEVEL_TRACE
+
+#ifndef AWS_LOG_LEVEL
+#define AWS_LOG_LEVEL AWS_DEFAULT_LOG_LEVEL
+#endif /* AWS_LOG_LEVEL */
+
+#define AWS_LOG(level, fmt, ...) \
+    do { \
+        AWS_WARNING_PUSH \
+        AWS_WARNING_DISABLE_CONST_CONDITIONAL \
+        if ((level) <= AWS_LOG_LEVEL) { \
+            aws_log((level), (fmt), ## __VA_ARGS__); \
+        } \
+        AWS_WARNING_POP \
+    } while (0)
+
+#define AWS_VLOG(level, fmt, va_args) \
+    do { \
+        AWS_WARNING_PUSH \
+        AWS_WARNING_DISABLE_CONST_CONDITIONAL \
+        if ((level) <= AWS_LOG_LEVEL) { \
+            aws_vlog((level), (fmt), (va_args)); \
+        } \
+        AWS_WARNING_POP \
+    } while (0)
+
+typedef void (aws_log_report_fn)(const char* log_message);
+
+
+#define AWS_LOG_MAX_LOGGING_THREADS (256)
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * `report_callback` is called from inside of `aws_log_flush`. Each message logged will be reported to this callback.
+ * The default reporting mechanism is an empty stub function (no-op).
+ */
+AWS_COMMON_API void aws_log_set_reporting_callback(aws_log_report_fn* report_callback);
+
+/**
+ * Initializes the logging system to capture future calls to `aws_log` and `aws_vlog`. The settings here are shared amongst
+ * all other instances of `aws_thread` when `aws_thread_launch` is called.
+ */
+AWS_COMMON_API int aws_log_system_init(struct aws_allocator *alloc, size_t max_message_len, int memory_pool_message_count, int max_message_count);
+
+/**
+ * Cleans up all logging related resources and flushes any remaining logs.
+ */
+AWS_COMMON_API void aws_log_system_clean_up();
+
+/**
+ * Records a log entry to be processed by a later call to `aws_log_flush`.
+ */
+AWS_COMMON_API int aws_log(enum aws_log_level level, const char *fmt, ...);
+
+/**
+ * Records a log entry to be processed by a later call to `aws_log_flush`.
+ */
+AWS_COMMON_API int aws_vlog(enum aws_log_level level, const char *fmt, va_list va_args);
+
+/**
+ * Returns the string representation of `level` as an `aws_log_level` enum type.
+ */
+AWS_COMMON_API const char *aws_log_level_to_string(enum aws_log_level level);
+
+/**
+ * Call this function to process any previous logs captured by `aws_log` calls. Can be called in a loop, on a condition variable,
+ * or by any other means deemed necessary.
+ */
+AWS_COMMON_API int aws_log_flush();

```

Comment:
Yes, you definitely can have a thread that blocks until some logs are available. The user would want to wrap `AWS_LOG` to receive a notification when logs appear. The user would then call `aws_log_process` from a thread waiting on a condition variable.

The question then becomes "can we implement this as apart of the logging API?" And the answer is definitely yes. But it's not implemented in this PR, as I didn't really see it as high of a priority as moving onto other things.

Reply:
Would it be a good idea for threads to call flush periodically, as a side-effect of calling aws_vlog()?

The implementation could be something like: aws_vlog() will flush if it's been X time since the last flush OR its memory pool is full. 

The user can still flush manually to prevent worker threads from doing the work, if they feel like.

Pros:
- Eliminates worrying about the user never calling flush
- Might let us put hard limits on the number of outstanding messages per thread.

Cons:
- Threads under load could be responsible for handling flush(), which increases their load.
- User is no longer under complete control of the thread that their reporting callback gets called on.

Reply:
We can't wrap AWS_LOG with something to kick our blocking thread as it's a preprocessor define that would apply to e.g. libraries we consume.

As for flushing within threads I assume this whole thing is meant to avoid blocking IO to flush logs in application work threads.